Sources/SwiftFusion/Inference/BearingRangeFactor.swift

```diff
@@ -0,0 +1,143 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import TensorFlow
+
+/// Protocol for indicating a struct is range-capable
+public protocol RangeFunction: Differentiable {
+  associatedtype Base: Differentiable
+  associatedtype Target: Differentiable
+  @differentiable
+  static func range(_ from: Base, _ to: Target) -> Double
+}
+
+public struct PointToPointRange2D: RangeFunction {
+  @differentiable
+  public static func range(_ from: Vector2, _ to: Vector2) -> Double {
+    (to - from).norm
+  }
+}
+
+/// Protocol for indicating a struct is bearing-capable
+public protocol BearingFunction: Differentiable {
+  associatedtype Base: Differentiable
+  associatedtype Target: Differentiable
+  associatedtype Bearing: Differentiable
+  @differentiable
+  static func bearing(_ from: Base, _ to: Target) -> Bearing
+}
+
+public struct PoseToPointBearingRange2D: RangeFunction & BearingFunction {
+  @differentiable
+  public static func bearing(_ from: Pose2, _ to: Vector2) -> Vector1 {
+    let dx = (to - from.t)
+    return Vector1(between(Rot2(c: dx.x / dx.norm, s: dx.y / dx.norm), from.rot).theta)
+  }
+  
+  @differentiable
+  public static func range(_ from: Pose2, _ to: Vector2) -> Double {
+    (to - from.t).norm
+  }
+}
+
+public struct BearingRangeError<BE: TangentStandardBasis & VectorConvertible & FixedDimensionVector>:
+  Differentiable & KeyPathIterable & TangentStandardBasis & VectorConvertible {
+  @differentiable
+  public init(_ vector: Vector) {
+    precondition(vector.dimension == BE.dimension + 1)
+    let part = vector.scalars.differentiablePartition(BE.dimension)
+    bearing = BE(Vector(part.a))
+    range = vector.scalars[BE.dimension]
+  }
+  
+  @differentiable
+  public init(bearing: BE, range: Double) {
+    self.bearing = bearing
+    self.range = range
+  }
+
+  @differentiable
+  public var vector: Vector {
+    get {
+      Vector(bearing.vector.scalars + [range])
+    }
+  }
+
+  public var bearing: BE
+  public var range: Double
+}
+
+/// A `NonlinearFactor` that calculates the bearing and range error of one pose and one landmark
+///
+/// Input is a dictionary of `Key` to `Value` pairs, and the output is the scalar
+/// error value
+///
+/// Interpretation
+/// ================
+/// `Input`: the input values as key-value pairs
+///
+public struct BearingRangeFactor<BearingRangeFunction: BearingFunction & RangeFunction>: NonlinearFactor
+where BearingRangeFunction.Base.TangentVector: VectorConvertible,
+      BearingRangeFunction.Target.TangentVector: VectorConvertible,
+      BearingRangeFunction.Bearing: Differentiable & TangentStandardBasis & EuclideanVector & VectorConvertible & FixedDimensionVector
+{
+  public typealias Base = BearingRangeFunction.Base
+  public typealias Target = BearingRangeFunction.Target
+  public typealias Bearing = BearingRangeFunction.Bearing
+  
+  var key1: Int
```

Comment:
Why int?

Reply:
Keys are still limited to `Int` in current SwiftFusion. Probably can be improved in the future with Marc's new FG backend.